File: Numerical_methods/0.cpp
#include <iostream>
#include <float.h>
#include <cmath>

using namespace std;

int main() {
    float eps;
    eps = 1;
    while (1 + eps > 1){
        eps = eps / 2;
    }
    cout << "Machine epsilon, float: " << eps * 2 << "\n";
    cout << "Theoretical machine epsilon, float: " << FLT_EPSILON << "\n\n";

    double eps2;
    eps2 = 1;
    while (1 + eps2 > 1) {
        eps2 = eps2 / 2;
    }
    cout << "Machine epsilon, double: " << eps2 * 2 << "\n";
    cout << "Theoretical machine epsilon, double: " << DBL_EPSILON << "\n\n";

    long double eps3;
    eps3 = 1;
    while (1 + eps3 > 1) {
        eps3 = eps3 / 2;
    }
    cout << "Machine epsilon, long double: " << eps3 * 2 << "\n";
    cout << "Theoretical machine epsilon, long double: " << LDBL_EPSILON << "\n-----------------------------------\n";


    float dwarf, real_dwarf;
    dwarf = 1;
    while (dwarf > 0){
        real_dwarf = dwarf;
        dwarf = dwarf / 2;
    }
    cout << "Machine dwarf, float: " << real_dwarf << "\n";
    cout << "Theoretical machine min, float: " << FLT_MIN << "\n\n";

    double dwarf2, real_dwarf2;
    dwarf2 = 1;
    while (dwarf2 > 0){
        real_dwarf2 = dwarf2;
        dwarf2 = dwarf2 / 2;
    }
    cout << "Machine dwarf, double: " << real_dwarf2 << "\n";
    cout << "Theoretical machine min, double: " << DBL_MIN << "\n\n";

    long double dwarf3, real_dwarf3;
    dwarf3 = 1;
    while (dwarf3 > 0){
        real_dwarf3 = dwarf3;
        dwarf3 = dwarf3 / 2;
    }
    cout << "Machine dwarf, long double: " << real_dwarf3 << "\n";
    cout << "Theoretical machine min, long double: " << LDBL_MIN << "\n-----------------------------------\n";

    float maximum, pre_maximum;
    maximum = 1;
    while (maximum != (float) INFINITY){
        pre_maximum = maximum;
        maximum = maximum * 2;
    }
    cout << "Maximum, float: " << pre_maximum << "\n";
    cout << "Theoretical maximum, float: " << FLT_MAX << "\n\n";

    double maximum2, pre_maximum2;
    maximum2 = 1;
    while (maximum2 != (double) INFINITY){
        pre_maximum2 = maximum2;
        maximum2 = maximum2 * 2;
    }
    cout << "Maximum, double: " << pre_maximum2 << "\n";
    cout << "Theoretical maximum, double: " << DBL_MAX << "\n\n";

    long double maximum3, pre_maximum3;
    maximum3 = 1;
    while (maximum3 != (long double) INFINITY){
        pre_maximum3 = maximum3;
        maximum3 = maximum3 * 2;
    }
    cout << "Maximum, long double: " << pre_maximum3 << "\n";
    cout << "Theoretical maximum, long double: " << LDBL_MAX << "\n\n";
    
}

File: Numerical_methods/0.py
import numpy as np

eps = np.half(1)
while np.half(1) + eps > np.half(1):
    eps /= np.half(2)
print("Machine epsilon, half:", eps * 2)
print("Theoretical machine epsilon, half:", np.finfo(np.half).eps, '\n')


eps2 = np.single(1)
while np.single(1) + eps2 > np.single(1):
    eps2 /= np.single(2)
print("Machine epsilon, single:", eps2 * 2)
print("Theoretical machine epsilon, single:", np.finfo(np.single).eps, '\n')


eps3 = np.double(1)
while np.double(1) + eps3 > np.double(1):
    eps3 /= np.double(2)
print("Machine epsilon, double:", eps3 * 2)
print("Theoretical machine epsilon, double:", np.finfo(np.double).eps, '\n')

print("--------------------------------------------------------------\n")

dwarf = np.half(1)
pre_dwarf = np.half(1)
while dwarf > np.half(0):
    pre_dwarf = dwarf
    dwarf /= np.half(2)
print("Machine dwarf, half:", pre_dwarf)
print("Theoretical machine dwarf, half:", np.finfo(np.half).tiny, '\n')

dwarf2 = np.single(1)
pre_dwarf2 = np.single(1)
while dwarf2 > np.single(0):
    pre_dwarf2 = dwarf2
    dwarf2 /= np.single(2)
print("Machine dwarf, single:", pre_dwarf2)
print("Theoretical machine dwarf, single:", np.finfo(np.single).tiny, '\n')

dwarf3 = np.double(1)
pre_dwarf3 = np.double(1)
while dwarf3 > np.double(0):
    pre_dwarf3 = dwarf3
    dwarf3 /= np.double(2)
print("Machine dwarf, double:", pre_dwarf3)
print("Theoretical machine dwarf, double:", np.finfo(np.double).tiny, '\n')

print("--------------------------------------------------------------\n")

maximum = np.half(1)
pre_maximum = np.half(1)
while maximum != np.half(np.inf):
    pre_maximum = maximum
    maximum *= np.half(2)
print("Machine maximum, half:", pre_maximum)
print("Theoretical machine maximum, half:", np.finfo(np.half).max, '\n')

maximum2 = np.single(1)
pre_maximum2 = np.single(1)
while maximum2 != np.single(np.inf):
    pre_maximum2 = maximum2
    maximum2 *= np.single(2)
print("Machine maximum, single:", pre_maximum2)
print("Theoretical machine maximum, single:", np.finfo(np.single).max, '\n')

maximum3 = np.double(1)
pre_maximum3 = np.double(1)
while maximum3 != np.double(np.inf):
    pre_maximum3 = maximum3
    maximum3 *= np.double(2)
print("Machine maximum, double:", pre_maximum3)
print("Theoretical machine maximum, double:", np.finfo(np.double).max, '\n')
